### ISSUE TITLE
feat: 8bit index

### ISSUE THREAD

**Issue by VoVAllen** [tensorchord/pgvecto.rs#196]
Use int8 or E5M2 to index vectors

**Comment by JoePassanante** [comment]
@VoVAllen I think this can be closed right?  With #170 done

**Comment by VoVAllen** [comment]
@JoePassanante We've done the fp16 index so far, but not int8 index yet. However, user can use Scalar quantization now which is very similar to 8bit index.

**Comment by my-vegetable-has-exploded** [comment]
Here is the how we use int8 to storage vector.
Given a vector of F32, [a_0, a_1, a_2, ..., a_n], we aim to find the maximum and minimum values. The maximum value, max, is the greatest among {a_0, a_1, a_2, ..., a_n}, and the minimum value, min, is the smallest.
We can transform F32 to I8 using the formula (a - min) / (max - min) * 255, resulting in a vector of I8, [b_0, b_1, b_2, ..., b_n]. Here 255 is the range size that the int8 type can cover, which is the difference between -128 and 127.

![image](https://github.com/tensorchord/pgvecto.rs/assets/48236141/f0f764dd-1363-4fd5-a80c-a0c9295a9beb)

But there are some problems,  as this article mentioned(https://qdrant.tech/articles/scalar-quantization/),  vectori8 is faster than  fp32. Those codes show than them use vnni ( i8*i8 -> i16 under simd, in short ) to speed up the process.
https://github.com/qdrant/quantization/blob/939fdb627a8edcf92fd71e3c79017156690850e9/quantization/cpp/avx2.c#L33-L46.  
I don't know whether the rust/llvm would auto-vectorize the process. If so , how to write the code without using raw simd instructions. 

By the way, have we ever write benchmarks to compare the performance between FP32 and UINT8 quantization ，，， since I think quantization also introduces some overhead? 

Another problem is that to if we want to use vnni to speed up the process, we need to introduce some precomputed results. How should we refactor the distance function in trait G,  since we need to pass more parameters.

**Comment by usamoi** [comment]
> I don't know whether the rust/llvm would auto-vectorize the process.

1. maybe we can have a try
2. pgvecto.rs uses `multiversion` for auto-vectorized versions of functions. You may search for its usage in the code.

How should we refactor the distance function in trait G, since we need to pass more parameters.
1. Add a new distance function to `G`. There is no good solution.

**Comment by VoVAllen** [comment]
What is the method of qdrant in your diagram?

> By the way, have we ever write benchmarks to compare the performance between FP32 and UINT8 quantization ，，， since I think quantization also introduces some overhead?

Based on my preliminary experiment, quantized by row seems have better results. https://colab.research.google.com/drive/1X_DZjZyy9HO7SrURJZiUzNsjt6muarBH?usp=sharing

> I don't know whether the rust/llvm would auto-vectorize the process.

Need experiments. But I think int8 should be supported

> Another problem is that to if we want to use vnni to speed up the process, we need to introduce some precomputed results. How should we refactor the distance function in trait G, since we need to pass more parameters.

Good question. I think we can keep the square sum of the elements, for both L2 and cos distance

**Comment by my-vegetable-has-exploded** [comment]
> What is the method of qdrant in your diagram?

Given a vector of F32 with shape(10000, 100), qdrant will split it into multi-parts like segments in out approach. For example, we take vector[0:1000] as first segment, it will calculate min/max value for the whole vector with shape(1000,100), and than use the min/max value to quantize. When a vector q query this segment, it will quantize q (this process need to be done only once for each segment. And it will use q'( q after quantization) to compute similarity. Since the min/max value is same for the segment, the computational cost will be reduced.

> > By the way, have we ever write benchmarks to compare the performance between FP32 and UINT8 quantization ，，， since I think quantization also introduces some overhead?
> 
> Based on my preliminary experiment, quantized by row seems have better results. https://colab.research.google.com/drive/1X_DZjZyy9HO7SrURJZiUzNsjt6muarBH?usp=sharing

I want to know that have we ever conducted a comparison in terms of computational efficiency?

**Comment by VoVAllen** [comment]
@my-vegetable-has-exploded No performance number yet. But 8bit should be much faster 32bit with proper AVX instruction

**Comment by my-vegetable-has-exploded** [comment]
> @my-vegetable-has-exploded No performance number yet. But 8bit should be much faster 32bit with proper AVX instruction

Sorry for the misunderstanding. I am asking if 8-bit scalar quantization (which we already implement) would  improved computational efficiency? 

https://github.com/tensorchord/pgvecto.rs/blob/c6dabc79ea3e0a47effb14c7a2b5ecbccbdff7af/crates/service/src/prelude/global/f32_cos.rs#L70-L71

The reason I'm concerned about this is that it introduces a cast from u8 back to f32 and still use f32 to compute the result.

**Comment by VoVAllen** [comment]
Shouldn't cast to f32 to compute. In SQ, the input vector is fp32, so the computation between i8 and fp32 might have better precision, at the cost of computational  efficiency. In i8 type, you only have the i8 representation. Use fp32 or i8 should have the same result, but i8 can be much faster
